compute_tools/src/bin/compute_ctl.rs

```diff
@@ -275,33 +277,86 @@ fn main() -> Result<()> {
         "running compute with features: {:?}",
         state.pspec.as_ref().unwrap().spec.features
     );
+    // before we release the mutex, fetch the swap size (if any) for later.
+    let swap_size_bytes = state.pspec.as_ref().unwrap().spec.swap_size_bytes;
     drop(state);
 
     // Launch remaining service threads
     let _monitor_handle = launch_monitor(&compute);
     let _configurator_handle = launch_configurator(&compute);
 
-    // Start Postgres
+    let mut prestartup_failed = false;
     let mut delay_exit = false;
-    let mut exit_code = None;
-    let pg = match compute.start_compute(extension_server_port) {
-        Ok(pg) => Some(pg),
-        Err(err) => {
-            error!("could not start the compute node: {:#}", err);
-            let mut state = compute.state.lock().unwrap();
-            state.error = Some(format!("{:?}", err));
-            state.status = ComputeStatus::Failed;
-            // Notify others that Postgres failed to start. In case of configuring the
-            // empty compute, it's likely that API handler is still waiting for compute
-            // state change. With this we will notify it that compute is in Failed state,
-            // so control plane will know about it earlier and record proper error instead
-            // of timeout.
-            compute.state_changed.notify_all();
-            drop(state); // unlock
-            delay_exit = true;
-            None
+
+    // Resize swap to the desired size if the compute spec says so
+    if let (Some(size_bytes), true) = (swap_size_bytes, resize_swap_on_bind) {
+        // TODO(cloud#12047/sharnoff): For avoid 'swapoff' hitting during postgres startup, we need
+        // to run resize-swap to completion *before* starting postgres.
+        //
+        // Maybe we want to do this asynchronously in the future? If so, we should probably set
+        // SkipSwapon for VMs before doing that (so that we're not running 'swapoff' during or
+        // after postgres startup).
+        // We may also *not* want to do it asynchronously, so we can guarantee that the swap is
+        // actually available before postgres starts (otherwise we may have potential issues).
+        // Or maybe the startup delay isn't worth it, and if postgres OOMs it'll restart
+        // successfully once the swap is added.
```

Comment:
Have we verified that it restarts on its own after OOM? Won't it tell the control plane that we failed to start compute?

Reply:
Hm... if postgres is OOM-killed _after_ startup it'll get restarted by compute_ctl (compute_ctl exits without delaying, and then is restarted externally, which then triggers starting postgres again).

But I guess it depends when it happens during startup — if `ComputeNode::start_compute` fails, then compute_ctl will go through the `delay_exit` path and wait for the control plane to notice that startup failed.

So maybe it won't get restarted, actually...

Reply:
Revised the comment here: 5f3f06f3978fecca4d38386dd15db18348a6474c

```
// To avoid 'swapoff' hitting postgres startup, we need to run resize-swap to completion
// *before* starting postgres.
//
// In theory, we could do this asynchronously if SkipSwapon was enabled for VMs, but this
// carries a risk of introducing hard-to-debug issues - e.g. if postgres sometimes gets
// OOM-killed during startup because swap wasn't available yet.
```